ssl/tls13_enc.c

```diff
@@ -0,0 +1,217 @@
+                             const unsigned char *hash,
+                             unsigned char *out, size_t outlen)
+{
+    const unsigned char label_prefix[] = "TLS 1.3, ";
+    const EVP_MD *md = ssl_handshake_md(s);
+    EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_id(EVP_PKEY_HKDF, NULL);
+    int ret;
+    size_t hkdflabellen;
+    size_t hashlen;
+    /*
+     * 2 bytes for length of whole HkdfLabel + 1 byte for length of combined
+     * prefix and label + bytes for the label itself + bytes for the hash
+     */
+    unsigned char hkdflabel[sizeof(uint16_t) + sizeof(uint8_t) +
+                            + sizeof(label_prefix) + TLS13_MAX_LABEL_LEN
+                            + EVP_MAX_MD_SIZE];
+    WPACKET pkt;
+
+    if (pctx == NULL)
+        return 0;
+
+    hashlen = EVP_MD_size(md);
+
+    if (!WPACKET_init_static_len(&pkt, hkdflabel, sizeof(hkdflabel), 0)
+            || !WPACKET_put_bytes_u16(&pkt, outlen)
+            || !WPACKET_start_sub_packet_u8(&pkt)
+            || !WPACKET_memcpy(&pkt, label_prefix, sizeof(label_prefix) - 1)
+            || !WPACKET_memcpy(&pkt, label, labellen)
+            || !WPACKET_close(&pkt)
+            || !WPACKET_sub_memcpy_u8(&pkt, hash, (hash == NULL) ? 0 : hashlen)
+            || !WPACKET_get_total_written(&pkt, &hkdflabellen)
+            || !WPACKET_finish(&pkt)) {
+        WPACKET_cleanup(&pkt);
+        return 0;
+    }
+
+    ret = EVP_PKEY_derive_init(pctx) <= 0
+            || EVP_PKEY_CTX_hkdf_mode(pctx, EVP_PKEY_HKDEF_MODE_EXPAND_ONLY)
+               <= 0
+            || EVP_PKEY_CTX_set_hkdf_md(pctx, md) <= 0
+            || EVP_PKEY_CTX_set1_hkdf_key(pctx, secret, hashlen) <= 0
+            || EVP_PKEY_CTX_add1_hkdf_info(pctx, hkdflabel, hkdflabellen) <= 0
+            || EVP_PKEY_derive(pctx, out, &outlen) <= 0;
+
+    EVP_PKEY_CTX_free(pctx);
+
+    return ret == 0;
+}
+
+/*
+ * Given a input secret |insecret| and a |label| of length |labellen|, derive a
+ * new |secret|. This will be the length of the current hash output size and
+ * will be based on the current state of the handshake hashes.
+ *
+ * Returns 1 on success  0 on failure.
+ */
+int tls13_derive_secret(SSL *s, const unsigned char *insecret,
+                        const unsigned char *label, size_t labellen,
+                        unsigned char *secret)
+{
+    unsigned char hash[EVP_MAX_MD_SIZE];
+    size_t hashlen;
+
+    if (!ssl3_digest_cached_records(s, 1))
+        return 0;
+
+    if (!ssl_handshake_hash(s, hash, sizeof(hash), &hashlen))
+        return 0;
+
+    return tls13_hkdf_expand(s, insecret, label, labellen, hash, secret,
+                             hashlen);
+}
+
+/*
+ * Given a |secret| generate a |key| of length |keylen| bytes.
+ *
+ * Returns 1 on success  0 on failure.
+ */
+int tls13_derive_key(SSL *s, const unsigned char *secret, unsigned char *key,
+                     size_t keylen)
+{
+    return tls13_hkdf_expand(s, secret, keylabel, sizeof(keylabel) - 1, NULL,
```

Comment:
I can't decide whether the ugliness of a "static const unsigned char keylabel[] = {'k', 'e', 'y'};" makes up for the ugliness of "sizeof(keylabel) - 1" here.
(Does the style guide have anything to say on this question?)